File: src/core/loss.ts
import { Tensor } from "./tensor";

export interface Loss {
  forward(pred: Tensor, target: Tensor): number;
  backward(pred: Tensor, target: Tensor): Tensor;
}

/**
 * Mean Squared Error loss.
 */
export class MSE implements Loss {
  forward(pred: Tensor, target: Tensor): number {
    let sum = 0;
    for (let i = 0; i < pred.data.length; i++) {
      const diff = pred.data[i] - target.data[i];
      sum += diff * diff;
    }
    return sum / pred.data.length;
  }

  backward(pred: Tensor, target: Tensor): Tensor {
    const grad = new Float32Array(pred.data.length);
    for (let i = 0; i < pred.data.length; i++) {
      grad[i] = 2 * (pred.data[i] - target.data[i]) / pred.data.length;
    }
    return new Tensor(grad, pred.shape);
  }
}

/**
 * Cross-Entropy loss with softmax.
 */
export class CrossEntropy implements Loss {
  forward(pred: Tensor, target: Tensor): number {
    // Assume pred is logits, target is one-hot or class indices
    // For simplicity, assume target is class indices, pred is [batch, classes]
    const batch = pred.shape[0];
    const classes = pred.shape[1];
    let loss = 0;
    for (let b = 0; b < batch; b++) {
      // Softmax
      let maxLogit = -Infinity;
      for (let c = 0; c < classes; c++) {
        maxLogit = Math.max(maxLogit, pred.data[b * classes + c]);
      }
      let sumExp = 0;
      for (let c = 0; c < classes; c++) {
        sumExp += Math.exp(pred.data[b * classes + c] - maxLogit);
      }
      const logSumExp = maxLogit + Math.log(sumExp);
      const targetClass = target.data[b];
      loss += logSumExp - pred.data[b * classes + targetClass];
    }
    return loss / batch;
  }

  backward(pred: Tensor, target: Tensor): Tensor {
    // Grad = softmax(pred) - target (one-hot)
    const batch = pred.shape[0];
    const classes = pred.shape[1];
    const grad = new Float32Array(pred.data.length);
    for (let b = 0; b < batch; b++) {
      // Softmax
      let maxLogit = -Infinity;
      for (let c = 0; c < classes; c++) {
        maxLogit = Math.max(maxLogit, pred.data[b * classes + c]);
      }
      let sumExp = 0;
      const exps = new Float32Array(classes);
      for (let c = 0; c < classes; c++) {
        exps[c] = Math.exp(pred.data[b * classes + c] - maxLogit);
        sumExp += exps[c];
      }
      for (let c = 0; c < classes; c++) {
        const softmax = exps[c] / sumExp;
        const targetVal = c === target.data[b] ? 1 : 0;
        grad[b * classes + c] = softmax - targetVal;
      }
    }
    return new Tensor(grad, pred.shape).mul(1 / batch);
  }
}

File: archive/train-lle-py/src/train_lle/loss.py
from .tensor import Tensor
import numpy as np

class Loss:
    def forward(self, pred, target):
        raise NotImplementedError

    def backward(self, pred, target):
        raise NotImplementedError

class MSE(Loss):
    def forward(self, pred, target):
        diff = pred.data - target.data
        return np.mean(diff ** 2)

    def backward(self, pred, target):
        diff = pred.data - target.data
        grad = 2 * diff / pred.data.size
        return Tensor(grad, pred.shape)

class CrossEntropy(Loss):
    def forward(self, pred, target):
        # Assume pred is logits [batch, classes], target is class indices [batch]
        batch, classes = pred.shape
        loss = 0
        for b in range(batch):
            logits = pred.data[b]
            max_logit = np.max(logits)
            log_sum_exp = max_logit + np.log(np.sum(np.exp(logits - max_logit)))
            target_class = int(target.data[b])
            loss += log_sum_exp - logits[target_class]
        return loss / batch

    def backward(self, pred, target):
        # Grad = softmax(pred) - one_hot(target)
        batch, classes = pred.shape
        grad = np.zeros_like(pred.data)
        for b in range(batch):
            logits = pred.data[b]
            max_logit = np.max(logits)
            exps = np.exp(logits - max_logit)
            softmax = exps / np.sum(exps)
            target_class = int(target.data[b])
            one_hot = np.zeros(classes)
            one_hot[target_class] = 1
            grad[b] = softmax - one_hot
        return Tensor(grad / batch, pred.shape)